File: scripts/check_locale_mix.py
#!/usr/bin/env python3
"""Dev-скрипт: ищет кириллицу в исходниках вне файлов локалей.

Назначение: видеть TODO по переводу — где ещё остался хардкод RU‑текста на бэке.
Не падает (exit 0): просто печатает список нарушений. В CI можно завернуть в must-fail.

Запуск:
    python -m scripts.check_locale_mix          # все файлы
    python -m scripts.check_locale_mix app/api  # отдельная директория

Конфиг:
    EXCLUDE_DIRS  — каталоги, которые игнорируем (venv, миграции, тесты, текстовые
                    словари локалей и moderation roots — там по дизайну есть RU).
    EXCLUDE_FILES — точечные исключения (брендовые тексты, словари локалей).
"""

from __future__ import annotations

import os
import re
import sys
from pathlib import Path

# Кириллический паттерн с обрамляющим контекстом (имя файла:строка:текст).
CYRILLIC_RE = re.compile(r"[А-яЁёА-я]+")

EXCLUDE_DIRS = {
    ".git", ".venv", "venv", "__pycache__", "node_modules",
    "static",
    # i18n‑словари — там кириллица по дизайну.
    os.path.join("app", "i18n"),
    # Бот‑тексты ещё переводим итеративно; не шумим.
    os.path.join("app", "texts"),
    # Дефолтные корни модерации (RU‑часть остаётся).
    # Сам файл проверяем, но эти константы пропускаем через EXCLUDE_FILES.
}

EXCLUDE_FILES = {
    # Эти модули содержат RU‑константы (словари модерации, бренды).
    os.path.join("app", "db", "ensure_defaults.py"),
    os.path.join("app", "moderation_lexicon.py"),
    os.path.join("scripts", "seed_profanity.py"),
    os.path.join("scripts", "check_locale_mix.py"),
    os.path.join("app", "handlers", "moderation.py"),
}

ALLOWED_EXTS = {".py"}


def _iter_files(root: Path):
    for path in root.rglob("*"):
        if path.is_dir():
            continue
        if path.suffix not in ALLOWED_EXTS:
            continue
        rel = str(path.relative_to(root))
        parts = set(rel.split(os.sep))
        # имя любой части пути совпадает с записью в EXCLUDE_DIRS
        if parts & {p for p in EXCLUDE_DIRS if os.sep not in p}:
            continue
        # точечный префикс ("app/i18n", "app/texts") — пропускаем целиком
        if any(rel.startswith(p + os.sep) or rel == p for p in EXCLUDE_DIRS if os.sep in p):
            continue
        if rel in EXCLUDE_FILES:
            continue
        yield path, rel


def _scan_file(path: Path, rel: str) -> list[tuple[int, str]]:
    out: list[tuple[int, str]] = []
    try:
        text = path.read_text(encoding="utf-8")
    except Exception:
        return out
    for i, line in enumerate(text.splitlines(), start=1):
        if CYRILLIC_RE.search(line):
            out.append((i, line.strip()[:180]))
    return out


def main(argv: list[str]) -> int:
    root = Path(argv[1] if len(argv) > 1 else ".").resolve()
    total = 0
    files_count = 0
    for path, rel in _iter_files(root):
        hits = _scan_file(path, rel)
        if not hits:
            continue
        files_count += 1
        print(f"\n# {rel} — RU mix ({len(hits)} line(s))")
        for ln, snippet in hits[:10]:
            print(f"  {ln}: {snippet}")
        if len(hits) > 10:
            print(f"  … +{len(hits) - 10} more")
        total += len(hits)
    print(f"\nSummary: {total} RU-mix line(s) in {files_count} file(s).")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: webapp/scripts/check_locale_mix.mjs
#!/usr/bin/env node
/**
 * Dev-скрипт: ищет кириллицу в исходниках Mini App вне локализационных файлов.
 *
 * Не падает (exit 0) — печатает оставшиеся хардкодные RU‑строки, чтобы видеть TODO.
 *
 * Запуск:
 *   node webapp/scripts/check_locale_mix.mjs [base-dir]
 */

import { readdir, readFile, stat } from 'node:fs/promises'
import { join, relative, sep } from 'node:path'
import process from 'node:process'

const CYRILLIC = /[А-яЁё]+/

const root = process.argv[2] ? process.argv[2] : process.cwd()

const EXCLUDE_DIRS = new Set([
  'node_modules',
  'dist',
  '.git',
  '.cache',
  '.vite',
  'i18n', // ru.js / en.js
])

const ALLOWED_EXT = new Set(['.vue', '.js', '.ts', '.jsx', '.tsx'])

const EXCLUDE_FILES = new Set([
  // Утилиты форматирования RU‑дат — оставляем как есть.
  ['src', 'utils', 'formatDateTime.js'].join(sep),
  ['scripts', 'check_locale_mix.mjs'].join(sep),
])

function shouldSkipDir(name, rel) {
  if (EXCLUDE_DIRS.has(name)) return true
  return false
}

async function* walk(dir) {
  let entries
  try {
    entries = await readdir(dir, { withFileTypes: true })
  } catch {
    return
  }
  for (const e of entries) {
    const full = join(dir, e.name)
    const rel = relative(root, full)
    if (e.isDirectory()) {
      if (shouldSkipDir(e.name, rel)) continue
      yield* walk(full)
      continue
    }
    if (!e.isFile()) continue
    const dot = e.name.lastIndexOf('.')
    if (dot === -1) continue
    const ext = e.name.slice(dot)
    if (!ALLOWED_EXT.has(ext)) continue
    if (EXCLUDE_FILES.has(rel)) continue
    yield { full, rel }
  }
}

async function scanFile(full) {
  let txt
  try {
    txt = await readFile(full, 'utf-8')
  } catch {
    return []
  }
  const lines = txt.split(/\r?\n/)
  const hits = []
  for (let i = 0; i < lines.length; i += 1) {
    const line = lines[i]
    if (CYRILLIC.test(line)) {
      hits.push({ ln: i + 1, snippet: line.trim().slice(0, 180) })
    }
  }
  return hits
}

async function main() {
  try {
    const stats = await stat(root)
    if (!stats.isDirectory()) {
      console.error(`Not a directory: ${root}`)
      process.exit(0)
    }
  } catch {
    console.error(`Not found: ${root}`)
    process.exit(0)
  }

  let total = 0
  let files = 0
  for await (const { full, rel } of walk(root)) {
    const hits = await scanFile(full)
    if (!hits.length) continue
    files += 1
    console.log(`\n# ${rel} — RU mix (${hits.length} line(s))`)
    for (const h of hits.slice(0, 10)) {
      console.log(`  ${h.ln}: ${h.snippet}`)
    }
    if (hits.length > 10) {
      console.log(`  … +${hits.length - 10} more`)
    }
    total += hits.length
  }
  console.log(`\nSummary: ${total} RU-mix line(s) in ${files} file(s).`)
}

main().catch((e) => {
  console.error(e)
  process.exit(0)
})
